src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -643,7 +660,163 @@ private static size_t CurlSendCallback(IntPtr buffer, size_t size, size_t nitems
+                // from that read result.  After that, we can transfer as much data remains between Offset and
+                // Count.  Multiple callbacks may pull from that one read.
+
+                EasyRequest.SendTransferState sts = easy._sendTransferState;
+                if (sts != null)
+                {
+                    // Is there a previous read that may still have data to be consumed?
+                    if (sts._task != null)
+                    {
+                        Debug.Assert(sts._task.IsCompleted, "The task must have completed if we're getting called back.");
+
+                        // Determine how many bytes were read on the last asynchronous read.
+                        // If nothing was read, then we're done and can simply return 0 to indicate
+                        // the end of the stream.
+                        int bytesRead = sts._task.GetAwaiter().GetResult(); // will throw if read failed
+                        Debug.Assert(bytesRead >= 0 && bytesRead <= sts._buffer.Length, "ReadAsync returned an invalid result length: " + bytesRead);
+                        if (bytesRead == 0)
+                        {
+                            multi.VerboseTrace("End of stream from stored task", easy: easy);
+                            sts.Update(null, 0, 0);
+                            return 0;
+                        }
+
+                        // If Count is still 0, then this is the first time after the task completed
+                        // that we're examining the data: transfer the bytesRead to the Count.
+                        if (sts._count == 0)
+                        {
+                            multi.VerboseTrace("ReadAsync completed with bytes: " + bytesRead, easy: easy);
+                            sts._count = bytesRead;
+                        }
+
+                        // Now Offset and Count are both accurate.  Determine how much data we can copy to libcurl...
+                        int availableData = sts._count - sts._offset;
+                        Debug.Assert(availableData > 0, "There must be some data still available.");
+
+                        // ... and copy as much of that as libcurl will allow.
+                        int bytesToCopy = Math.Min(availableData, length);
+                        Marshal.Copy(sts._buffer, sts._offset, buffer, bytesToCopy);
+                        multi.VerboseTrace("Copied " + bytesRead + " bytes from request stream", easy: easy);
+
+                        // Update the offset.  If we've gone through all of the data, reset the state 
+                        // so that the next time we're called back we'll do a new read.
+                        sts._offset += bytesToCopy;
+                        Debug.Assert(sts._offset <= sts._count, "Offset should never exceed count");
+                        if (sts._offset == sts._count)
```

Comment:
I think that'd only make sense if the update function was the only way those fields were updated.  I really meant for this type to be a simple POD, and the helper function just made for a concise way of updating all three fields.  If the helper function to set the three fields is confusing things, I'd prefer to remove it than make it more complex.  For now, I've just renamed it to SetTaskOffsetCount to make it very clear what it's doing.